src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,214 @@
+      rw [C_mul_X_pow_of_card_support_le_one Cf, C_mul_X_pow_of_card_support_le_one Cg],
+      rw [mul_assoc, X_pow_mul, mul_assoc, ← pow_add X],
+      repeat {rw rp_smul},
+      rw [rp_mon _ _ Nf, rp_mon _ _ Og, rp_mon],
+      { rw [mul_assoc, X_pow_mul, mul_assoc, ← pow_add X],
+        congr,
+        rcases (nat.le.dest Og) with ⟨ G, rfl ⟩,
+        rcases (nat.le.dest Nf) with ⟨ F, rfl ⟩,
+        repeat {rw nat.add_sub_cancel_left},
+        rw [add_comm _ F, add_comm G F, add_assoc, nat.add_sub_assoc, ← add_assoc],
+        rw [add_comm f.nat_degree _, add_comm (_+f.nat_degree) G, nat.add_sub_cancel],
+        rw add_comm,
+        exact add_le_add_left Og f.nat_degree, },
+      { rw add_comm N O,
+        exact add_le_add Og Nf, }, },
+    -- second induction: induction step
+    { intros N O f g Cf Cg Nf Og,
+      by_cases g0 : g = 0,
+      { rw [g0, mul_zero, rp_zero, rp_zero, mul_zero], },
+      { rw [← erase_lead_add_C_mul_X_pow g, mul_add, rp_add, rp_add, mul_add],
+        rw hcg N O f _ Cf _ Nf (le_trans erase_lead_nat_degree_le Og),
+        rw hcg N O f _ Cf (le_add_left card_support_C_mul_X_pow_le_one) Nf _,
+        { exact (le_trans (nat_degree_C_mul_X_pow_le g.leading_coeff g.nat_degree) Og), },
+        { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cg (erase_lead_support_card_lt g0)), }, }, }, },
+  --first induction: induction step
+  { intros N O f g Cf Cg Nf Og,
+    by_cases f0 : f=0,
+    { rw [f0, zero_mul, rp_zero, rp_zero, zero_mul], },
+    { rw [← erase_lead_add_C_mul_X_pow f, add_mul, rp_add, rp_add, add_mul],
+      rw hcf N O _ g _ Cg (le_trans erase_lead_nat_degree_le Nf) Og,
+      rw hcf N O _ g (le_add_left card_support_C_mul_X_pow_le_one) Cg _ Og,
+      { exact (le_trans (nat_degree_C_mul_X_pow_le f.leading_coeff f.nat_degree) Nf), },
+      { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cf (erase_lead_support_card_lt f0)), }, }, },
+end
+
+lemma pol_ind_Rhom_prod {rp : ℕ → polynomial R → polynomial R}
+ (rp_add  : ∀ f g : polynomial R, ∀ F : ℕ,
+  rp F (f+g) = rp F f + rp F g)
+ (rp_smul : ∀ f : polynomial R, ∀ r : R, ∀ F : ℕ,
+  rp F ((C r)*f) = C r * rp F f)
+ (rp_mon : ∀ N n : ℕ, n ≤ N →
+  rp N (X^n) = X^(N-n)) :
+ ∀ N O : ℕ, ∀ f g : polynomial R,
+ f.nat_degree ≤ N → g.nat_degree ≤ O →
+ (rp (N + O) (f*g)) = (rp N f) * (rp O g) :=
+begin
+  intros _ _ f g,
+  apply pol_ind_Rhom_prod_on_card f.support.card g.support.card rp_add rp_smul rp_mon,
+  { exact (f.support).card.le_succ, },
+  { exact (g.support).card.le_succ, },
+end
+
+@[simp] theorem reflect_mul
+ {f g : polynomial R} {F G : ℕ} (Ff : f.nat_degree ≤ F) (Gg : g.nat_degree ≤ G) :
+ reflect (F+G) (f*g) = reflect F f * reflect G g :=
+begin
+  apply pol_ind_Rhom_prod reflect_add reflect_smul _ F G f g Ff Gg,
+  { intros N n Nn,
+    rw [reflect_monomial, rev_at_small Nn], },
+end
+
+theorem reverse_mul (f g : polynomial R) {fg : f.leading_coeff*g.leading_coeff ≠ 0} :
+ reverse (f*g) = reverse f * reverse g :=
```

Comment:
Lean can't come up with `fg` by looking at the terms in the conclusion, so we have to make it explicit. Conversely, `f` and `g` appear in `fg` so we can make them implicit:
```suggestion
theorem reverse_mul {f g : polynomial R} (fg : f.leading_coeff * g.leading_coeff ≠ 0) :
 reverse (f * g) = reverse f * reverse g :=
```